subsys/net/lib/lwm2m/lwm2m_engine.c

```diff
@@ -2354,10 +2407,59 @@ void lwm2m_udp_receive(struct lwm2m_ctx *client_ctx, struct net_pkt *pkt,
 		if (handle_separate_response && !tkl &&
 			zoap_header_get_type(&response) == ZOAP_TYPE_ACK) {
 			SYS_LOG_DBG("separated response, not removing reply");
+			goto cleanup;
+		}
+	}
+
+	if (reply || pending) {
+		/* free up msg resources */
+		if (msg) {
+			lwm2m_release_message(msg);
+		}
+
+		SYS_LOG_DBG("reply %p handled and removed", reply);
+		goto cleanup;
+	}
+
+	/*
+	 * If no normal response handler is found, then this is
+	 * a new request coming from the server.  Let's look
+	 * at registered objects to find a handler.
+	 */
+	if (udp_request_handler &&
+	    zoap_header_get_type(&response) == ZOAP_TYPE_CON) {
+		msg = lwm2m_get_message(client_ctx);
+		if (!msg) {
+			SYS_LOG_ERR("Unable to get a lwm2m message!");
+			goto cleanup;
+		}
+
+		/* Create a response message if we reach this point */
+		msg->type = ZOAP_TYPE_ACK;
+		msg->code = zoap_header_get_code(&response);
+		msg->mid = zoap_header_get_id(&response);
+		/* tkl -1 means don't set the token at all */
+		msg->tkl = -1;
```

Comment:
tkl is type u8_t and cannot be negative.

Reply:
TKL is a 4-bit unsigned integer (per RFC 7252), we could use MSB for not generating a new token.

Reply:
or use 65535 as dont generate (what -1 equates to as a u8_t )

Reply:
Yup